### ISSUE TITLE
[MH] rejection-free proposal for the multivariate normal case

### ISSUE THREAD

**Issue by emilemathieu** [yebai/Turing.jl#387]
This [blog post](http://fsaad.scripts.mit.edu/randomseed/metropolis-hastings-sampling-with-gaussian-drift-proposal-on-bounded-support/) shows how to sample from a truncated (univariate) gaussian proposal for Metropolis-Hasting, so as to avoid rejecting a MH step because the proposed value does not lie on the distribution's support.
This is currently [implemented in Turing](https://github.com/yebai/Turing.jl/pull/369/files#diff-901e29b3cdf668ffac2c0cfec047bdaaR199).

Can we extend this idea to the normal multivariate case ?

**Comment by ae-foster** [comment]
Yes, we can extend to the multivariate case. But the kinds of constraints possible is much richer. This is why, even thought `Distributions.jl` has a truncated Gaussian, it doesn't have a truncated Multivariate gaussian (see https://github.com/JuliaStats/Distributions.jl/issues/480). They couldn't agree what a 'typical' constraint was.

For our case, rectangular constraints are the simplest. It's just a product of many univariate interval constraints. The correction factor would then be the product of the correction factors for each dimension.

A linear constraint that doesn't align with the univariate distributions is also fine. (A constraint of the form a <= X.r <= b.) This is again easy because X.r is univariate Gaussian so the correction factors are again made up of Normal cdf values.

A third constraint type you might be interested in is ||X|| <= a. Again, it is possible to calculate the correction factor though you need to do the eigendecomposition of the covariance matrix. (See https://mathoverflow.net/questions/155829/euclidian-norm-of-gaussian-vectors).

So basically, what kinds of constraints are we interested in supporting?

**Comment by emilemathieu** [comment]
I think there is a misunderstanding, the issue is not in sampling from a truncated gaussian (even if that's interesting to know how to do that without sampling until constraints are satisfied) but to know how to correct the acceptance ratio of the Metropolis-Hasting algorithm.

**Comment by ae-foster** [comment]
Yes @emilemathieu , but the correction factors depend on the type of constraint!

**Comment by ae-foster** [comment]
Let me slow down. Suppose we have general constraint set C. Then the correction ratio is made up of objects that look like P(X \in C) where X is a multivariate Gaussian. We may or may not be able to compute that probability analytically. It all depends upon the shape of C

**Comment by emilemathieu** [comment]
True ! For instance, the constraint could be the set of positive semi-definite matrix, for a MH step on a covariance matrix with a Normal proposal.

**Comment by ae-foster** [comment]
So you draw the matrix components from a multivariate Gaussian and subsequently constraint the resulting matrix to be positive definite?

**Comment by emilemathieu** [comment]
Hum I think I said a mistake. But we could have a simplex constraint for instance.

**Comment by ae-foster** [comment]
Sure, we could have many constraints! I guess I'm asking which are the important ones? Which ones are so common that we cannot tolerate regular rejection?

**Comment by emilemathieu** [comment]
That's a good point, I'm trying to think of common constraint on vector random variable.

**Comment by emilemathieu** [comment]
I'm not sure it's worth spending too much time on this, maybe we should just close it.

**Comment by ae-foster** [comment]
Ok let's close this. What we should do is this. If we constantly find ourselves doing MH with variables on a simplex (say) and we notice that proposals keep getting rejected, then we open a new issue asking for the correction factors for *that type* of constraint. Until then, we might be optimizing prematurely by writing a corrected proposer which never gets used.
